Update capabilities to payment methods mapping

Mayisha opened the pull request:
## Changes proposed in this Pull Request:
- We are already filtering out the payment methods in the backend which do not have capabilities in the Stripe account. Removed the duplicate filtering from frontend in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3375/commits/6dc17143187ad768f6b69fbd13847948df39469f
- If the payment method was enabled in Stripe dashboard, it might have different key in the `capabilities` object. For example `klarna` instead of `klarna_payments`. With the changes in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3375/commits/d7b4cede298664d0ad249491e4b065e130e5d621 we are considering both type of keys (`pm` and `pm_payments`) to check availability of the payment methods in the corresponding Stripe account.

## Testing instructions
- Connect to your Stripe account.
- Enable `live` mode.
- From `network` tab, check the capabilities of your account from `<your_site>/wp-json/wc/v3/wc_stripe/account`.
- Ensure that the payment methods that are not present in your `capabilities` object, are hidden from the payment methods list.
- Enable `test` mode and confirm that all the payment methods are displayed irrespective of their `capabilities` status.

#### Note
Asked here p1724345677895689/1724279352.583359-slack-C06RFP01Z7G to confirm about the correct key for `card`method.

Mayisha committed:
update capability to payment method mapping
- consider `pm` and `pm_payments` both type of keys to check capabilities

Mayisha committed:
remove capability check from frontend
- payment methods are already filtered in the backend in 'filter_payment_methods_with_capabilities' method

james-allan reviewed (approved):
Changes look good. I confirmed that while in live mode, payment methods that I don't have the capability for (eg Ali Pay) weren't shown.

In test mode all payment method types were available. 

I wasn't able to confirm the capability format difference. All of my account's capabilities were in the format `{type}_payments`, but I trust that Stripe's advice is good. :)

james-allan commented:
@Mayisha I just noticed there's a changelog missing.

Mayisha committed:
Merge branch 'develop' into update/payment-method-mapping-to-capabilities

Mayisha committed:
add changelog entry

Mayisha committed:
consider `legacy_payments` capability for card method.

Mayisha committed:
refactor and add comment

Mayisha commented:
Thank you @james-allan.

I have added the changelog in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3375/commits/e75871eb6db181a9a8f21483b96db0bedf118d84

Apart from that, I did a little refactoring to consider the legacy card capability (`legacy_payments`) in the mapping. and also added some more comments in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3375/commits/1d5ee0b8f70e60c077a30cdb05723c3972ef3617 and [3e045ba](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3375/commits/3e045ba60c8a3ba7f67ee76c86d756acb6e43503)

james-allan commented on includes/class-wc-stripe-helper.php:
```diff
@@ -726,12 +726,21 @@ public static function filter_payment_methods_with_capabilities( $payment_method
+		// Return all payment methods if in test mode.
+		if ( $testmode ) {
+			return $payment_method_ids;
+		}
```
Oh good change. 👍 Skipping the whole process in test mode makes so much more sense 😂

james-allan commented on includes/class-wc-stripe-helper.php:
```diff
@@ -726,12 +726,21 @@ public static function filter_payment_methods_with_capabilities( $payment_method
+			// Generally the key is the payment method id appended with '_payments' (i.e. 'card_payments', 'sepa_debit_payments', 'klarna_payments').
+			// In some cases, the Stripe account might have the legacy key set. For example, for Klarna, the legacy key is 'klarna'.
+			// For card, the legacy key is 'legacy_payments'.
+			$has_capability = isset( $data['capabilities'][ $key ] ) || isset( $data['capabilities'][ $payment_method_id ] ) || ( 'card' === $payment_method_id && $data['capabilities']['legacy_payments'] );
```
```suggestion
			$has_capability = isset( $data['capabilities'][ $key ] ) || isset( $data['capabilities'][ $payment_method_id ] ) || ( 'card' === $payment_method_id && isset( $data['capabilities']['legacy_payments'] ) );
```

There's a missing isset check on the final `legacy_payments`, if it ever got to that condition it would error if it wasn't set. 

```
PHP Warning:  Undefined array key "legacy_payments" in /wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-helper.php on line 746

Mayisha replied:
That's a good catch @james-allan 😬 
Fixed in [132cbca](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3375/commits/132cbcafeec9379a8267f734e5c49aa716937d43)

Mayisha committed:
Merge branch 'develop' into update/payment-method-mapping-to-capabilities

Mayisha committed:
isset check

james-allan reviewed (approved):
Changes look good. Thanks @Mayisha!

Mayisha committed:
Merge branch 'develop' into update/payment-method-mapping-to-capabilities
